server/api/assets.js

```diff
@@ -1,14 +1,164 @@
+      throw new ValidationError('wallet address not match');
+    }
+
+    // check ts expire
+    if (Math.abs(ts - Date.now()) > ONE_DATE_IN_MS) {
+      throw new ValidationError('payload expired');
+    }
+
+    // check tag
+    if (tags.length < 1) throw new ValidationError('asset need at least 1 tag');
+    if (tags.length > MAX_TAG_COUNT) {
+      throw new ValidationError(`asset need at most ${MAX_TAG_COUNT} tag`);
+    }
+    const isTagsValid = tags.every(e => e.length >= MIN_TAG_LENGTH && e.length <= MAX_TAG_LENGTH);
+    if (!isTagsValid) throw new ValidationError('tags length invalid');
+
+    // check license
+    if (!license) throw new ValidationError('asset need a license');
+    if (!SUPPORTED_LICENSE.has(license)) throw new ValidationError('license invalid');
+    
+    // check asset
+    const { file: asset } = req;
+    if (!asset) throw new ValidationError('no asset uploaded');
+    if (asset.size > MAX_IMAGE_SIZE) throw new ValidationError('asset size too large');
+    const type = imageType(asset.buffer);
+    if (!SUPPORTED_IMAGE_TYPE.has(type && type.ext)) throw new ValidationError('unsupported file format!');
+    const hash256 = sha256(asset.buffer);
+    if (hash256 !== assetSHA256) throw new ValidationError('asset sha not match');
+    const hash256Bytes = Buffer.from(hash256, 'hex');
+    const oldAsset = await sequelize.asset.findById(hash256Bytes, { raw: true });
+    if (oldAsset) throw new ValidationError('asset already exist');
+
+    // FIXME: use ipfs wrapper function
+    const ipfs = await IpfsClient();
+    const ipfsAdd = await ipfs.files.add(asset.buffer);
+    if (!ipfsAdd) throw new Error('fail to ipfs add');
+    await ipfs.pin.add(ipfsAdd[0].hash);
+
+    // metadata
+    // TODO: creator: get user id by wallet address
+    const licenseObj = await sequelize.license.findById(license, { raw: true });
+    const uploadDate = new Date().toISOString();
+    const mediaObj = {
+      creator: [wallet],
+      dateCreated,
+      description,
+      license: licenseObj.url,
+      fingerprint: assetSHA256,
+      footprint: footprint || [],
+      ipfs: ipfsAdd[0].hash,
+      previousVersion: null,
+      uploadDate,
+    };
+    const formattedMediaObj = formatMediaObject(mediaObj);
+    const ipld = await ipfs.dag.put(formattedMediaObj, { format: 'dag-cbor', hashAlg: 'sha2-256' });
+    await ipfs.pin.add(ipld.toBaseEncodedString());
+
+    // db update
+    // NOTE: sequelize.model.bulkCreate fail on duplicate in postgres
+    await sequelize.sequelize.query(`INSERT INTO tag (name) VALUES
+            ${tags.map(tag => `('${tag}')`).join(', ')}
+            ON CONFLICT DO NOTHING;`);
+    const assetRecord = {
+      fingerprint: hash256Bytes,
+      ipfs: bs58.decode(ipfsAdd[0].hash),
+      ipld: bs58.decode(ipld.toBaseEncodedString()),
+      wallet,
+      fk_asset_license: license,
+    };
+    await sequelize.asset.upsert(assetRecord);
```

Comment:
Insert and should throw asset already exist error.

Reply:
http://docs.sequelizejs.com/class/lib/model.js~Model.html
I found that there is no insert method. And using upsert is an issue because it will update the values on duplicate exception. Is writing a raw query to insert only solution?

Reply:
Model.create() is insert method

http://docs.sequelizejs.com/manual/tutorial/instances.html